proposals/0000-complete-set-signatures.rst

```diff
@@ -0,0 +1,483 @@
+constraints of a pattern snyonym.)
+
+Examples
+--------
+
+The example from the introduction:
+
+::
+
+ class IsList l where
+   viewCons :: l a -> Maybe (a, l a)
+
+ pattern (:<) :: IsList l => a -> l a -> l a
+ pattern x :< xs <- (viewCons -> Just (x,xs))
+ pattern Empty :: IsList l => l a
+ pattern Empty <- (viewCons -> Nothing)
+
+ {-# COMPLETE (:<), Empty :: IsList l => l a #-} -- (1)
+
+ instance IsList [] where ...
+
+ f :: IsList l => l Int -> Int
+ f Empty    = 0
+ f (x :< _) = x
+
+ class IsList l => IsInfinite l where
+   -- INVARIANT: `viewCons` always returns `Just`
+
+ data Stream a = S a (Stream a)
+ instance IsList Stream where
+   viewCons (S x xs) = Just (x, xs)
+ instance IsInfinite Stream where
+
+ {-# COMPLETE (:<) :: IsInfinite l => l a #-} -- (2)
+
+ instance IsInfinite Stream where ...
+
+ safeHead :: IsInfinite l => l a -> a
+ safeHead (x :< _) = x
+
+ safeHead2 :: Stream a -> a
+ safeHead2 (x :< _) = x
+
+ unsafeHead :: [a] -> a
+ unsafeHead (x :< _) = x
+
+This program passes type-checking. The compiler *should* emit a warning about
+the definition of ``unsafeHead`` being incomplete, but not for ``f``,
+``safeHead`` or ``safeHead2``:
+
+- ``f`` has a case for ``Empty`` and ``(:<)``. COMPLETE set (1) is covered,
+  because the type of the pattern match is ``IsList l => l a``, which is
+  subsumed by itself. Thus, the pattern match of ``f`` is exhaustive.
+- ``f`` has a case for ``(:<)``. COMPLETE set (2) is *not* covered,
+  because the type of the pattern match is ``IsList l => l a``, which is not
+  subsumed by ``IsInfinite l => l a``.
+- ``safeHead`` has a case for ``(:<)``. COMPLETE set (2) is covered,
+  because the type of the pattern match is ``IsInfinite l => l a``, which is
+  subsumed by itself. Thus, the pattern match of ``safeHead`` is exhaustive.
+- ``safeHead2`` has a case for ``(:<)``. COMPLETE set (2) is covered,
+  because the type of the pattern match is ``Stream a``, which is
+  subsumed by ``IsInfinite l => l a``.
+  Thus, the pattern match of ``safeHead2`` is exhaustive.
+- ``unsafeHead`` has a case for ``(:<)``. COMPLETE set (2) is *not* covered,
+  because the type of the pattern match is ``[a]``, which is not
+  subsumed by ``IsInfinite l => l a``.
+- The lack of any COMPLETE set being covered by the the pattern match in
+  ``unsafeHead`` means that its definition is flagged as inexhaustive.
+
+Constraints in the signature
+============================
+
+Given the following example (full example in `#14422 <https://gitlab.haskell.org/ghc/ghc/-/issues/14422#note_296728>`_) ::
+
+ pattern FZ :: () =>        n ~ 'S m      => Fin n
+ pattern FZ <- (viewFin -> VZ) where
+   FZ = Fin 0
+
+ pattern FS :: () => n ~ 'S m => Fin m -> Fin n
+ pattern FS m <- (viewFin -> VS m) where
+   FS (Fin m) = Fin (1 + m)
+
+ {-# COMPLETE FZ, FS #-}
+
+ inc :: Fin n -> Fin (S n)
+ inc FZ       = FS FZ
+ inc n@(FS _) = FS n
+
+How would I give the signature a COMPLETE set while preserving its semantics?
```

Comment:
I didn't understand this entire paragraph.  Reword/clarify?

Reply:
This was a word mixup. I meant

> What result type signature would I give the COMPLETE set that preserves its semantics?

I hope that is clearer.